R/di-fit.R

```diff
@@ -0,0 +1,276 @@
+# -----------------------------------------------------------------------------
+
+apd_di_bridge <- function(training, validation, importance, ...) {
+  blueprint <- training$blueprint
+  training <- training$predictors
+  all_numeric <- purrr::map_lgl(training, is.numeric)
+
+  if (!is.null(validation)) {
+    validation <- validation$predictors
+    all_numeric <- c(all_numeric, purrr::map_lgl(validation, is.numeric))
+  }
+
+  all_numeric <- all(all_numeric)
+
+  if (!all_numeric) {
+    rlang::abort(
+      "All predictors must be numeric",
+      call = rlang::caller_env()
+    )
+  }
+
+  all_importance <- all(names(training) %in% importance[["Variable"]])
+
+  if (!all_importance) {
+    rlang::abort(
+      "All predictors must have an importance value in `importance`",
+      call = rlang::caller_env()
+    )
+  }
+
+  fit <- apd_di_impl(training, validation, importance, ...)
+
+  new_apd_di(
+    training = fit$training,
+    importance = fit$importance,
+    sds = fit$sds,
+    means = fit$means,
+    d_bar = fit$d_bar,
+    aoa_threshold = fit$aoa_threshold,
+    blueprint = blueprint
+  )
+}
+
+# -----------------------------------------------------------------------------
+# ----------------------- Model function interface ----------------------------
+# -----------------------------------------------------------------------------
+
+#' Fit a `apd_di`
+#'
+#' `apd_di()` fits a model.
+#'
+#' @param x,y Depending on the context:
+#'
+#'   * A __data frame__ of predictors.
+#'   * A __matrix__ of predictors.
+#'   * A __recipe__ specifying a set of preprocessing steps
+#'     created from [recipes::recipe()].
+#'
+#'  If `y` is `NULL`, then this function will use within-sample distances.
+#'  This may result in too large an area of applicability being calculated.
+#'
+#' @param data,validation When a __recipe__ or __formula__ is used, `data` is specified as:
+#'
+#'   * A __data frame__ containing the predictors.
+#'
+#'  If `validation` is `NULL`, then this function will use within-sample distances.
+#'  This may result in too large an area of applicability being calculated.
+#'
+#' @param importance A data.frame with two columns: `Variable`, containing
+#' the names of each variable in the training and validation data, and
+#' `Importance`, containing the (raw or scaled) feature importance for each
+#' variable.
+#'
+#' @param formula A formula specifying the predictor terms on the right-hand
+#' side.
+#'
+#' @param ... Not currently used, but required for extensibility.
+#'
+#' @details This function calculates the "area of applicability" of a model, as
+#' introduced by Meyer and Pebesma (2021). While the initial paper introducing
+#' this method focused on spatial models, there is nothing inherently spatial
+#' about the method; it can be used with any type of data.
+#'
+#' The `importance` argument is structured to work with objects returned by the
```

Comment:
We should define some sort of tidy format consistent with broom (at least in the naming convention). I think columns  `term` and `estimate` (or maybe `value`) would be good. 

Also, we need to differentiate between a zero importance due to scaling (like caret does) or due to not being used in the model. Also also, we might want to require all of the variables to be included. 

Can you define a straw man specification for that?

Reply:
Maybe I missed it but is there any validation code for the concordance of the values in the importance df and the columns of the data being passed? 

For example, how would dummy variables be dealt with between the two objects, etc?

Reply:
I'm not sure that we can do anything about this but what happens if there is a difference in data processing between the model and the APD method? So, let's say a model uses a recipe to preprocess/transform the data. If the recipe doesn't match that, will that cause an error or poor results? 

It would be good to say something about that in the documentation (as a reminder).